File: apps/patter-voice-sdk/libraries/python/getpatter/engines/openai.py
"""OpenAI Realtime engine marker for Patter."""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    # Imported for the ``turn_detection`` forward-reference annotation only.
    # ``RealtimeTurnDetection`` validates itself in its own ``__post_init__``
    # (models.py) so no construction-time validation is needed on the marker —
    # the value is already validated before it reaches us.
    from getpatter.models import RealtimeTurnDetection

__all__ = ["Realtime"]


@dataclass(frozen=True)
class Realtime:
    """OpenAI Realtime API engine config.

    Holds the minimal settings needed by the Patter server to instantiate
    :class:`getpatter.providers.openai_realtime.OpenAIRealtimeAdapter` at call time.

    Example::

        from getpatter.engines import openai

        engine = openai.Realtime()                     # reads OPENAI_API_KEY
        engine = openai.Realtime(voice="nova", model="gpt-4o-mini-realtime-preview")
        engine = openai.Realtime(
            model="gpt-realtime-2",
            reasoning_effort="low",                     # gpt-realtime-2 only
            input_audio_transcription_model="gpt-realtime-whisper",
        )
    """

    api_key: str = ""
    voice: str = "alloy"
    model: str = "gpt-realtime-mini"
    # Reasoning-effort tier for ``gpt-realtime-2``. ``None`` leaves the
    # ``session.reasoning`` field unset (server default applies). OpenAI
    # recommends ``"low"`` for production voice flows — higher tiers add
    # measurable per-turn latency. No effect on models that ignore the field.
    reasoning_effort: Literal["minimal", "low", "medium", "high"] | None = None
    # Override for the Realtime session's ``input_audio_transcription.model``.
    # ``None`` keeps the adapter default (``whisper-1``). Use
    # ``"gpt-realtime-whisper"`` for low-latency partials, ``"gpt-4o-transcribe"``
    # for higher accuracy.
    input_audio_transcription_model: str | None = None
    # Input noise reduction for speakerphone / conference audio. ``None``
    # (default) omits the field (no reduction). ``"far_field"`` recommended
    # for phone calls. Mirrors ``Patter.agent(openai_realtime_noise_reduction=)``.
    noise_reduction: Literal["near_field", "far_field"] | None = None
    # Turn-detection tuning (raise the VAD threshold to reject speakerphone
    # noise, or switch to ``semantic_vad`` eagerness='low'). ``None`` (default)
    # keeps the adapter's current turn_detection.
    turn_detection: "RealtimeTurnDetection | None" = None
    # When ``True``, gate the model response on the Whisper transcript
    # arriving (legacy behavior). ``None``/``False`` (default) decouples the
    # response: the model replies as soon as the user stops speaking
    # (``input_audio_buffer.committed``), so it no longer waits ~500 ms for
    # Whisper. The transcript becomes pure observability (dashboard / history /
    # ``on_transcript``). Mirrors ``Patter.agent(realtime_gate_response_on_transcript=)``.
    gate_response_on_transcript: bool | None = None

    def __post_init__(self) -> None:
        key = self.api_key or os.environ.get("OPENAI_API_KEY", "")
        if not key:
            raise ValueError(
                "OpenAI Realtime engine requires an api_key. Pass "
                "api_key='sk-...' or set OPENAI_API_KEY in the environment."
            )
        object.__setattr__(self, "api_key", key)

    @property
    def kind(self) -> str:
        """Stable discriminator used for Phase 2 dispatch."""
        return "openai_realtime"

File: apps/patter-voice-sdk/libraries/typescript/src/engines/openai.ts
/** OpenAI Realtime engine — marker class for Patter client dispatch. */

import type { RealtimeTurnDetection } from '../types';
import { validateRealtimeTurnDetection } from '../providers/openai-realtime';

/** Constructor options for the OpenAI `Realtime` engine marker. */
export interface RealtimeOptions {
  /** API key. Falls back to OPENAI_API_KEY env var when omitted. */
  apiKey?: string;
  /**
   * Realtime model. Defaults to ``gpt-realtime-mini`` (bumped from the
   * deprecated ``gpt-4o-mini-realtime-preview`` on 2026-05-25 for
   * parity with the Python SDK and the GA Realtime API surface).
   */
  model?: string;
  /** Voice preset. Defaults to alloy. */
  voice?: string;
  /**
   * Reasoning-effort tier for `gpt-realtime-2`. When omitted the
   * `session.reasoning` field is not sent and the server default applies.
   * OpenAI recommends `"low"` for production voice flows — higher tiers add
   * measurable per-turn latency. Has no effect on models that ignore the
   * field.
   */
  reasoningEffort?: 'minimal' | 'low' | 'medium' | 'high';
  /**
   * Override for the Realtime session's `input_audio_transcription.model`.
   * Omit to keep the adapter default (`whisper-1`). Use
   * `"gpt-realtime-whisper"` for low-latency transcript partials,
   * `"gpt-4o-transcribe"` for higher accuracy.
   */
  inputAudioTranscriptionModel?: string;
  /**
   * Input noise reduction for speakerphone / conference audio. `undefined`
   * (default) omits the field (no reduction). `"far_field"` recommended for
   * phone / speakerphone calls; `"near_field"` for a handset close to the
   * mouth. Mirrors `openai_realtime_noise_reduction` on `Patter.agent()`.
   */
  noiseReduction?: 'near_field' | 'far_field';
  /**
   * Turn-detection tuning. `undefined` (default) keeps the adapter's
   * current hardcoded `server_vad` / threshold `0.5` / silence 300 ms.
   * Raise threshold or switch to `semantic_vad` eagerness `'low'` to stop
   * speakerphone noise from triggering false barge-ins.
   *
   * Maps to `turn_detection` on the Python `engines.openai.Realtime` marker;
   * propagates to `realtimeTurnDetection` on `AgentOptions`.
   */
  turnDetection?: RealtimeTurnDetection;
  /**
   * Gate the model's response on the Whisper transcript (legacy behavior).
   *
   * `false` (default) — the speech-to-speech model responds as soon as the
   * user stops speaking (on `speech_stopped`), independently of the Whisper
   * input transcription. The transcript becomes a pure observability
   * side-channel (dashboard / history / `onTranscript`) and never gates,
   * triggers, or cancels the response. This reclaims ~500 ms of latency
   * because the model no longer waits for Whisper.
   *
   * `true` — restores the prior behavior where the response is requested
   * only after the Whisper `transcript_input` event arrives and passes the
   * hallucination filter.
   *
   * Maps to `gate_response_on_transcript` on the Python
   * `engines.openai.Realtime` marker; propagates to
   * `openaiRealtimeGateResponseOnTranscript` on `AgentOptions`.
   */
  gateResponseOnTranscript?: boolean;
}

/**
 * OpenAI Realtime engine marker.
 *
 * @example
 * ```ts
 * import * as openai from "getpatter/engines/openai";
 * const engine = new openai.Realtime();                     // reads OPENAI_API_KEY
 * const engine = new openai.Realtime({ voice: "alloy" });
 * const engine = new openai.Realtime({
 *   model: "gpt-realtime-2",
 *   reasoningEffort: "low",                                  // gpt-realtime-2 only
 *   inputAudioTranscriptionModel: "gpt-realtime-whisper",
 * });
 * ```
 */
export class Realtime {
  readonly kind = "openai_realtime" as const;
  readonly apiKey: string;
  readonly model: string;
  readonly voice: string;
  readonly reasoningEffort?: 'minimal' | 'low' | 'medium' | 'high';
  readonly inputAudioTranscriptionModel?: string;
  readonly noiseReduction?: 'near_field' | 'far_field';
  readonly turnDetection?: RealtimeTurnDetection;
  readonly gateResponseOnTranscript?: boolean;

  constructor(opts: RealtimeOptions = {}) {
    const key = opts.apiKey ?? process.env.OPENAI_API_KEY;
    if (!key) {
      throw new Error(
        "OpenAI Realtime requires an apiKey. Pass { apiKey: 'sk-...' } or " +
          "set OPENAI_API_KEY in the environment.",
      );
    }
    if (opts.noiseReduction !== undefined && opts.noiseReduction !== 'near_field' && opts.noiseReduction !== 'far_field') {
      throw new Error(
        `noiseReduction must be 'near_field' or 'far_field', got ${JSON.stringify(opts.noiseReduction)}`,
      );
    }
    validateRealtimeTurnDetection(opts.turnDetection);
    this.apiKey = key;
    this.model = opts.model ?? "gpt-realtime-mini";
    this.voice = opts.voice ?? "alloy";
    this.reasoningEffort = opts.reasoningEffort;
    this.inputAudioTranscriptionModel = opts.inputAudioTranscriptionModel;
    this.noiseReduction = opts.noiseReduction;
    this.turnDetection = opts.turnDetection;
    this.gateResponseOnTranscript = opts.gateResponseOnTranscript;
  }
}
